go/listbasic.go

```diff
@@ -0,0 +1,153 @@
+package bucketclient
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"strconv"
+)
+
+type ListBasicOption func(*listBasicOptionSet) error
+
+// ListBasicGTOption only lists keys greater than the given argument
+func ListBasicGTOption(gt string) ListBasicOption {
+	return func(opts *listBasicOptionSet) error {
+		opts.gt = gt
+		return nil
+	}
+}
+
+// ListBasicGTEOption only lists keys greater or equal to the given argument
+func ListBasicGTEOption(gte string) ListBasicOption {
+	return func(opts *listBasicOptionSet) error {
+		opts.gte = gte
+		return nil
+	}
+}
+
+// ListBasicLTOption only lists keys less than the given argument
+func ListBasicLTOption(lt string) ListBasicOption {
+	return func(opts *listBasicOptionSet) error {
+		opts.lt = lt
+		return nil
+	}
+}
+
+// ListBasicLTEOption only lists keys less or equal to the given argument
+func ListBasicLTEOption(lte string) ListBasicOption {
+	return func(opts *listBasicOptionSet) error {
+		opts.lte = lte
+		return nil
+	}
+}
+
+// ListBasicMaxKeysOption limits the number of returned keys (default and maximum is 10000).
+func ListBasicMaxKeysOption(maxKeys int) ListBasicOption {
+	return func(opts *listBasicOptionSet) error {
+		if maxKeys < 0 || maxKeys > 10000 {
+			return fmt.Errorf("maxKeys=%d is out of the valid range [0, 10000]", maxKeys)
+		}
+		opts.maxKeys = maxKeys
+		return nil
+	}
+}
+
+// ListBasicNoKeysOption declares that keys are not needed in the
+// result entries and may be returned empty.
+//
+// Note: keys may still be returned until ARSN-438 is fixed.
+func ListBasicNoKeysOption() ListBasicOption {
+	return func(opts *listBasicOptionSet) error {
+		opts.noKeys = true
+		return nil
+	}
+}
+
+// ListBasicNoValuesOption declares that values are not needed in the
+// result entries and may be returned empty.
+//
+// Note: values may still be returned until ARSN-438 is fixed.
+func ListBasicNoValuesOption() ListBasicOption {
+	return func(opts *listBasicOptionSet) error {
+		opts.noValues = true
+		return nil
+	}
+}
+
+type ListBasicEntry struct {
+	Key   string `json:"key"`
+	Value string `json:"value"`
+}
+
+type ListBasicResponse []ListBasicEntry
+
+type listBasicOptionSet struct {
```

Comment:
Nitpicky: I think in the spirit of Go, you would use references here (instead of `string` and `int`) and check `== nil` later. This also takes care of the issue of `maxKeys` having to be initialized as `-1`.

Reply:
I agree as well, this will also avoid magic numbers like `-1` which have little to no meaning for a developer without documentation/context/reading code.
Adding to what @fredmnl mentioned, in golang, every type has a zero value (0 for integers, "" for strings, false for booleans, and nil for pointers). Relying on -1 as a special value conflicts with this zero value concept.